### PR TITLE
Migrate to Bzlmod

### PR DESCRIPTION
**Pull request by jadenPete**
This PR removes `WORKSPACE` and `tests/WORKSPACE`, migrating the ruleset to Bzlmod. For the most part, the ruleset can be used just as it was before (although including it is much simpler, thanks to Bzlmod's system for importing external repositories). There are a few breaking changes, though:
- Toolchains must be registered manually (i.e. without macros like `scala_register_toolchains` and `scala_proto_register_toolchains`), since Bzlmod only allows them to be registered in `MODULE.bazel`-like files
  - See https://bazel.build/external/migration#register-toolchains
- The default Scala toolchain is no longer set via `scala_register_toolchains`, since that macro no longer exists. Instead, users of the ruleset must set a flag like `--@rules_scala_annex//rules/scala:scala-toolchain=annex_zinc_2_13` in their `.bazelrc` file.
  - Previously, `scala_register_toolchains` invoked a repository rule that would generate the configuration value for the Scala toolchain. That way, the default could be set dynamically in the repository importing `rules_scala_annex`. This is no longer possible, since `rules_scala_annex` can't share repositories like `rules_scala_annex_scala_toolchain` with the repository that imports it.
- For the targets in `rules/scala_with_scalafmt.bzl`, the default Scalafmt configuration file can't be set via `scalafmt_default_config` for the same reason mentioned above. Instead, we set it via a toolchain and use the new `scalafmt_toolchain_name` attribute to override the file for specific targets.
- The `emulate_rules_scala` workspace macro, which was used to provide compatibility with `bazelbuild/rules_scala`'s rules, has been deleted. It hasn't been updated in years and relied on deprecated Bazel constructs like `bind` and `//external/...` that are no longer supported in Bzlmod. I figured that now that we're pursuing unifying the rulesets, there's little point in keeping it around.

### CODE REVIEW

File: MODULE.bazel [hunk 1]
```diff
@@ -1,6 +1,146 @@
...
-# Please consider migrating your external dependencies from WORKSPACE to MODULE.bazel.
-#
-# For more details, please check https://github.com/bazelbuild/bazel/issues/18958
-###############################################################################
+module(name = "rules_scala_annex")
+
+bazel_dep(name = "bazel_skylib", version = "1.7.1")
+
+# TODO: Use buildifier instead of buildifier_prebuilt once
+# https://github.com/bazelbuild/bazel-central-registry/issues/380 is resolved
```
Comment (jadenPete) [inline MODULE.bazel]: I realized after I wrote this that we could just use `archive_override`, but I don't think it makes much sense to have users of the ruleset fetch an entire Go toolchain and compile buildifier from scratch when it's only used internally.

File: MODULE.bazel [hunk 2]
```diff
@@ -1,6 +1,146 @@
...
+bazel_dep(name = "buildifier_prebuilt", version = "7.3.1")
```
Comment (jjudd) [inline MODULE.bazel]: Can we mark dev dependencies as such, so they don't propagate to downstream projects? I believe we can just set `dev_dependency = True` on the `bazel_dep` call to make that the case.

File: MODULE.bazel [hunk 3]
```diff
@@ -1,6 +1,146 @@
...
+bazel_dep(name = "rules_java", version = "7.12.2")
+bazel_dep(name = "rules_jvm_external", version = "6.5")
+bazel_dep(name = "rules_pkg", version = "1.0.1")
+bazel_dep(name = "stardoc", version = "0.7.1")
```
Comment (jjudd) [inline MODULE.bazel]: I think this is also a dev dependency

File: MODULE.bazel [hunk 4]
```diff
@@ -1,6 +1,146 @@
...
+module(name = "rules_scala_annex")
+
+bazel_dep(name = "bazel_skylib", version = "1.7.1")
+
+# TODO: Use buildifier instead of buildifier_prebuilt once
+# https://github.com/bazelbuild/bazel-central-registry/issues/380 is resolved
+bazel_dep(name = "buildifier_prebuilt", version = "7.3.1")
+bazel_dep(name = "rules_java", version = "7.12.2")
+bazel_dep(name = "rules_jvm_external", version = "6.5")
+bazel_dep(name = "rules_pkg", version = "1.0.1")
```
Comment (jjudd) [inline MODULE.bazel]: I think this is also a dev dependency

File: README.md [hunk 5]
```diff
@@ -54,81 +54,29 @@ straightforward.
...
+bazel_dep(name = "rules_scala_annex")
 
-# Load rules_scala_annex
-rules_scala_annex_version = "lucid_2024-11-18"
+rules_scala_annex_version = "<COMMIT>"
 
-http_archive(
-    name = "rules_scala_annex",
-    integrity = "sha256-WjZvojiclkiyVxQ1NqkH1lDeGaDLyzQOGiDsCfhVAec=",
+archive_override(
```
Comment (jjudd) [inline README.md:69]: Do folks need to set up toolchains in their repos with bzlmod or are our toolchains just set up in their repo by including our ruleset?

File: docs/newdocs/scala_versions.md [hunk 6]
```diff
@@ -19,64 +19,58 @@ unused/undeclared dependency checking and test code coverage checking) via the
...
     "register_zinc_toolchain",
 )
 
+
+# You'll need to pull these in via `rules_jvm_external`
 compiler_classpath_2_13 = [
-    "@scala_compiler_2_13//jar",
-    "@scala_library_2_13//jar",
-    "@scala_reflect_2_13//jar",
+    "@annex//:org_scala_lang_scala_compiler",
```
Comment (jjudd) [inline docs/newdocs/scala_versions.md]: Do we want these to be `@annex` or should we assume that the person setting up the Scala toolchain has their own maven dependency tree?

File: rules/scala/versions.bzl [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# Please ensure these stay up-to-date with the versions in `MODULE.bazel` and `tests/MODULE.bazel`.
+# Unfortunately, `MODULE.bazel` files can't call `load`, so we have to copy them there.
+scala_2_12_version = "2.12.19"
```
Comment (jjudd) [inline rules/scala/versions.bzl:4]: I don't love this and am not aware of any way around it. 

Is there any way to load variables out of the `MODULE.bazel` file?

File: tests/MODULE.bazel [hunk 8]
```diff
@@ -1,6 +1,156 @@
...
+
+zinc_version = "1.10.4"
+
+specs2_version = "4.20.9"
+
+scalatest_version = "3.2.19"
+
+scalacheck_version = "1.18.1"
+
+annex_test = use_extension("@rules_jvm_external//:extensions.bzl", "maven")
```
Comment (jjudd) [inline tests/MODULE.bazel:47]: I think for tests we may be able to use `include` because folks won't be depending on the test workspace: https://bazel.build/rules/lib/globals/module#include

Not sure if we should, but it is a thought.

File: tests/MODULE.bazel [hunk 9]
```diff
@@ -1,6 +1,156 @@
-###############################################################################
-# Bazel now uses Bzlmod by default to manage external dependencies.
-# Please consider migrating your external dependencies from WORKSPACE to MODULE.bazel.
-#
-# For more details, please check https://github.com/bazelbuild/bazel/issues/18958
-###############################################################################
+module(name = "rules_scala_annex_tests")
+
+bazel_dep(name = "bazel_skylib", version = "1.7.1")
```
Comment (jjudd) [inline tests/MODULE.bazel]: I think we can mark some of these as `dev_dependency`, but I'm not sure it matters here because no one's going to depend on the test workspace.

File: tests/MODULE.bazel [hunk 10]
```diff
@@ -1,6 +1,156 @@
...
+    "//scala:test_bootstrap_2_13",
+    "//scala:test_bootstrap_3",
+    "//scala:test_zinc_2_12",
+    "//scala:test_zinc_2_13",
+    "//scala:test_zinc_3",
+    "//scala:test_zinc_2_13_fatal_deprecation_opt",
+    "//scala:test_zinc_2_13_log_level_debug",
+)
+
+# Please ensure these stay up-to-date with the versions in `rules/scala/versions.bzl`.
```
Comment (jjudd) [inline tests/MODULE.bazel]: Should the path here be updated to refer to the parent workspace instead of the test workspace?

File: tests/coverage/BUILD [hunk 11]
```diff
@@ -8,14 +8,14 @@ scala_test(
     ],
     scala_toolchain_name = "test_zinc_2_13",
     runtime_deps = [
-        "@hamcrest_core//jar",
-        "@junit_interface//jar",
+        "@annex_test//:com_novocode_junit_interface",
```
Comment (jjudd) [inline tests/coverage/BUILD:11]: Nice does this mean you removed the last vestiges of the jvm imported deps and everything is now using rules_jvm_external? If so, that's awesome.

File: tests/MODULE.bazel [hunk 12]
```diff
@@ -1,6 +1,156 @@
...
+    artifact = "compiler-bridge_2.12",
+    group = "org.scala-sbt",
+    neverlink = True,
+    version = zinc_version,
+)
+use_repo(annex_test_2_12, "annex_test_2_12")
+
+http_file = use_repo_rule("@bazel_tools//tools/build_defs/repo:http.bzl", "http_file")
+
+http_file(
```
Comment (jjudd) [inline tests/MODULE.bazel:147]: This is scope creep, but can this be downloaded with rules_jvm_external instead?

File: tests/MODULE.bazel [hunk 13]
```diff
@@ -1,6 +1,156 @@
...
+    name = "shapeless_jar",
+    downloaded_file_path = "shapeless.jar",
+    sha256 = "072c83eca9996aed92310dc7225cfc313edc7b74a3a96e2bf25459ebfc04ac96",
+    urls = ["https://repo.maven.apache.org/maven2/com/chuusai/shapeless_2.13/2.3.7/shapeless_2.13-2.3.7.jar"],
+)
+
+http_file(
```
Comment (jjudd) [inline tests/MODULE.bazel:154]: Similar thought here.  Can this be downloaded with rules_jvm_external instead?

File: tests/dependencies/indirect/BUILD [hunk 14]
```diff
@@ -1,19 +1,19 @@
 load("@rules_scala_annex//rules:register_toolchain.bzl", "register_zinc_toolchain")
 load("@rules_scala_annex//rules:scala.bzl", "scala_library")
-load("@rules_scala_annex//rules/scala:workspace.bzl", "scala_2_13_version")
+load("@rules_scala_annex//rules/scala:versions.bzl", "scala_2_13_version")
```
Comment (jjudd) [inline tests/dependencies/indirect/BUILD:3]: It feels weird to have versions come from annex and then the jars come from the test workspace, but perhaps that is just how things are now?

File: MODULE.bazel [hunk 15]
```diff
@@ -1,6 +1,146 @@
...
+    "//:annex_scalafmt",
+    "//:repository_default_toolchain_21_definition",
+    "//rules/scala_proto:scalapb_scala_proto_toolchain",
+    "//src/main/scala:annex_bootstrap_2_13",
+    "//src/main/scala:annex_bootstrap_3",
+    "//src/main/scala:annex_zinc_2_13",
+    "//src/main/scala:annex_zinc_3",
+)
+
+# Please ensure these stay up-to-date with the versions in `rules/scala/versions.bzl`.
```
Comment (jjudd) [inline MODULE.bazel]: Can we update this to ask folks to keep these in sync with versions.bzl and the test MODULE.bazel?

File: tests/MODULE.bazel (hunk 10 above)
Comment (jjudd) [reply tests/MODULE.bazel]: Can we also update this comment to ask people to update the versions in annex's MODULE.bazel?

**Review by jjudd — COMMENTED** [review]
Took a look through things. Had a few questions and comments. Mostly LGTM.

File: MODULE.bazel (hunk 2 above)
Comment (jadenPete) [reply MODULE.bazel]: I originally tried marking this as a dev dependency, but then the test repository failed to build because we reference `buildifier_prebuilt` in `@rules_scala_annex//:BUILD.bazel`.

File: MODULE.bazel (hunk 3 above)
Comment (jadenPete) [reply MODULE.bazel]: See my comment here:
https://github.com/lucidsoftware/rules_scala/pull/74#discussion_r1873510258

File: MODULE.bazel (hunk 4 above)
Comment (jadenPete) [reply MODULE.bazel]: See my comment here:
https://github.com/lucidsoftware/rules_scala/pull/74#discussion_r1873510258

File: README.md (hunk 5 above)
Comment (jadenPete) [reply README.md:69]: Nope, they're set up by including our ruleset. I think that's the one aspect of external repositories in Bzlmod that isn't isolated—toolchains are shared throughout the entire Bazel server. See the [documentation](https://bazel.build/extending/toolchains#toolchain-resolution) on toolchain resolution, namely the bit about the order in which toolchains are resolved.

File: docs/newdocs/scala_versions.md (hunk 6 above)
Comment (jadenPete) [reply docs/newdocs/scala_versions.md]: Oh, that's a good point. The comment above says "You'll need to pull these in via `rules_jvm_external`", but it would be helpful to mention that `annex` should be replaced with the name of their dependency tree. I'll do that.

File: rules/scala/versions.bzl (hunk 7 above)
Comment (jadenPete) [reply rules/scala/versions.bzl:4]: I don't like it either and searched around for quite a while for a solution. Here's an issue I came across:
https://github.com/bazelbuild/bazel/issues/17880

Notably:
> I don't think I'm exaggerating that much with the following claim: The availability of and reliance on `load` in `WORKSPACE` was a major contributor to its subpar performance.

There is [`include`](https://bazel.build/rules/lib/globals/module#include), which allows `MODULE.bazel` files to be segmented, but the notable limitation of `include` is that "variable bindings (such as those used for `use_extension`) are only ever visible in the file they occur in, not in any included or including files".

I also considered writing a custom extension that would create the dependency trees and `load` these versions from `scala/versions.bzl`, but extensions can't call other extensions like `@rules_jvm_external//:extensions.bzl%maven` (i.e. extension macros don't exist), so this isn't an option either.

File: tests/MODULE.bazel (hunk 8 above)
Comment (jadenPete) [reply tests/MODULE.bazel:47]: Unfortunately, we can't `include` an `annex.MODULE.bazel` file to use `@annex` in tests because, according to the [documentation](https://bazel.build/rules/lib/globals/module#include), "only files in the main repo may be included".

I considered symlinking an `tests/annex.MODULE.bazel` file to `annex.MODULE.bazel`, but then `rules_scala_annex` would have to `include` it as well. That doesn't work because in the context of the `rules_scala_annex`-`rules_scala_annex_test` relationship, `rules_scala_annex` is not the root module (it's a dependency), and "Only the root module may use `include()`; it is an error if a `bazel_dep`'s `MODULE` file uses `include()`".

Honestly, though, now that I think about it, `rules_scala_annex_test` using `@annex` was a bad idea to begin with for a couple reasons:
- It's probably best for testing if it uses `rules_scala_annex` like any other repository would, which means using a separate dependency tree
- Its dependencies are different from `rules_scala_annex`; this is what we have `@annex_test` for

File: tests/MODULE.bazel (hunk 9 above)
Comment (jadenPete) [reply tests/MODULE.bazel]: We could, but yeah, it wouldn't benefit us at all. And if someone did depend on `rules_scala_annex_tests`, it would be incorrect for the reason I describe here:
https://github.com/lucidsoftware/rules_scala/pull/74#discussion_r1873510258

File: tests/MODULE.bazel (hunk 10 above)
Comment (jadenPete) [reply tests/MODULE.bazel]: > Should the path here be updated to refer to the parent workspace instead of the test workspace?

Doesn't `rules/scala/versions.bzl` imply it's in the main workspace because `tests/rules` doesn't exist? I could make it a relative path if you'd like

> Can we also update this comment to ask people to update the versions in annex's MODULE.bazel?

I figured that rather than having every set of variables point to all the others (`O(N^2)`), we could have every set point to `versions.bzl` and have that point to all the others (`O(N)`).

File: tests/coverage/BUILD (hunk 11 above)
Comment (jadenPete) [reply tests/coverage/BUILD:11]: It does!

File: tests/MODULE.bazel (hunk 13 above)
Comment (jadenPete) [reply tests/MODULE.bazel:154]: I considered migrating it to `rules_jvm_external`, but thought it was best if it continued to be fetched via `http_file`. That way, it wouldn't have `JavaInfo` or any extra providers added, which tests that `scala_import` doesn't rely on any of this information.

File: tests/dependencies/indirect/BUILD (hunk 14 above)
Comment (jadenPete) [reply tests/dependencies/indirect/BUILD:3]: Hmm, it does. Would you prefer we split `versions.bzl` into two files: one for `rules_scala_annex` and another for `rules_scala_annex_test`? I figured that even though we're forced to declare the versions in multiple places and define separate dependencies trees, it would still be best to have a single source of truth.

File: MODULE.bazel (hunk 15 above)
Comment (jadenPete) [reply MODULE.bazel]: See my comment here:
https://github.com/lucidsoftware/rules_scala/pull/74#discussion_r1873537315

File: MODULE.bazel (hunk 2 above)
Comment (jjudd) [reply MODULE.bazel]: Huh that's interesting. Do we need to do that?

File: README.md (hunk 5 above)
Comment (jjudd) [reply README.md:69]: Nice.

File: docs/newdocs/scala_versions.md (hunk 6 above)
Comment (jjudd) [reply docs/newdocs/scala_versions.md]: I suggest you rename it something like `@maven` I worry that people will miss your note and be confused as to how this is supposed to come from annex. I think both the `@maven` and your note will hopefully avoid that confusion.

File: rules/scala/versions.bzl (hunk 7 above)
Comment (jjudd) [reply rules/scala/versions.bzl:4]: I came to the same conclusion after doing some Googling. Nice to have a single source of truth for rule sets while also being able to `include` for end users.

I just don't love that there isn't a way to load something from the `MODULE.bazel` file into the rest of the workspace, e.g., the scala version.

File: tests/MODULE.bazel (hunk 8 above)
Comment (jjudd) [reply tests/MODULE.bazel:47]: Fair. I like this argument.

File: tests/MODULE.bazel (hunk 10 above)
Comment (jjudd) [reply tests/MODULE.bazel]: > Doesn't rules/scala/versions.bzl imply it's in the main workspace because tests/rules doesn't exist?
Only if I have context about the ruleset. If I didn't have context about the main rule set I'd not understand.

I also see your argument about the links to everything. That said, I really hope we only ever have 3 places where we're storing these versions, so I think it'd worth it to be more explicit here.

File: tests/MODULE.bazel (hunk 13 above)
Comment (jjudd) [reply tests/MODULE.bazel:154]: Ah. Do you want to add a comment to these `http_file`s explaining why they aren't part of `rules_jvm_external`? I think others may have the same thought as me in the future.

File: tests/dependencies/indirect/BUILD (hunk 14 above)
Comment (jjudd) [reply tests/dependencies/indirect/BUILD:3]: Yeah probably best to minimize the number of places versions can come from and leave it as is.

File: MODULE.bazel (hunk 15 above)
Comment (jjudd) [reply MODULE.bazel]: Similar response.

File: MODULE.bazel (hunk 2 above)
Comment (jadenPete) [reply MODULE.bazel]: We do for `BUILD` file formatting. I think we do it because the CI hosts don't necessarily have `buildifier` installed, so we need to pull it in through the repository.

File: docs/newdocs/scala_versions.md (hunk 6 above)
Comment (jadenPete) [reply docs/newdocs/scala_versions.md]: Good idea. I'll rename it.

File: rules/scala/versions.bzl (hunk 7 above)
Comment (jadenPete) [reply rules/scala/versions.bzl:4]: Agreed.

File: tests/MODULE.bazel (hunk 10 above)
Comment (jadenPete) [reply tests/MODULE.bazel]: That's fair. I'll update them to reference each other. I'll also prefix the paths to the files in which these are declared with `/` so it's clear they're all relative to the main workspace.

File: tests/MODULE.bazel (hunk 12 above)
Comment (jadenPete) [reply tests/MODULE.bazel:147]: See my comment here:
https://github.com/lucidsoftware/rules_scala/pull/74#discussion_r1873539489

File: tests/MODULE.bazel (hunk 13 above)
Comment (jadenPete) [reply tests/MODULE.bazel:154]: Will do.

**Comment by jadenPete** [comment]
Squashed.

**Review by jjudd — APPROVED** [review]
